website/server/controllers/api-v3/hall.js

```diff
@@ -163,16 +163,29 @@ api.getHero = {
   middlewares: [authWithHeaders(), ensureAdmin],
   async handler (req, res) {
     let heroId = req.params.heroId;
+    let validationErrors;
+    let hero;
 
-    req.checkParams('heroId', res.t('heroIdRequired')).notEmpty().isUUID();
+    req.checkParams('heroId', res.t('heroIdRequired')).notEmpty();
 
-    let validationErrors = req.validationErrors();
+    validationErrors = req.validationErrors();
     if (validationErrors) throw validationErrors;
 
-    let hero = await User
-      .findById(heroId)
-      .select(heroAdminFields)
-      .exec();
+    req.checkParams('heroId', res.t('heroIdRequired')).isUUID();
```

Comment:
I think the double validationError checking seems to be a bit off.

you could keep the first one with `notEmpty` and then test the param with `validator.isUUID()` without using `validationErrors()` again

Reply:
agreed

Reply:
Sorry I'm not really familar with the methods here...

`    if (req.checkParams('heroId', res.t('heroIdRequired')).isUUID() === true) {
      hero = await User
        .findById(heroId)
        .select(heroAdminFields)
        .exec();
    } else {
      hero = await User
        .findOne({'auth.local.username': heroId})
        .select(heroAdminFields)
        .exec();
    }`
By doing this, the code path always goes to the else block. And thus searching by user id will not work.

Reply:
I think what @negue is suggesting is that `req.validationErrors();` is only called once, so the second check can be done with the `validator` library, something like this:

```
import validator from 'validator';
...
req.checkParams('heroId', res.t('heroIdRequired')).notEmpty();
const validationErrors = req.validationErrors();
if (validationErrors) throw validationErrors; 
....
const heroParam = req.params.heroId;

let query;
if (validator.isUUID(heroParam)) {
    query = {'_id': heroParam};
} else {
    query = 'auth.local.username': heroId};
}

await User.findOne(query).select(...).exec()
...
```

it should work but I haven't tested this code so there are likely errors

Reply:
It seems like `.isUUID()` does not return `true` or `false`

Reply:
what does it return? if you search for `validator.isUUID` in the code you'll see it's used in many places and it returns true or false

Reply:
Ahh okay I tried it. I was confused because `params.isUUID()` did not return `true` or `false`, but I think the result is given in the `.validationErrors()` instead. 

`validator.isUUID()` returns true or false.